File: 3473.pas
const	maxn=100+5;maxm=10+5;mo=1000000007;len=17;size=(1 shl 5);jw=100000000000000000;
type	bignum=array[0..15] of int64;
	matrix=array[0..size,0..size] of int64;
var	i,j,m,n,tot,tmp:longint;
	st:array[0..(1 shl maxm),0..1] of longint;
	dp:array[0..maxn,0..(1 shl maxm)] of int64;
	s,nn,mm:string;						//change
	base,ans:matrix;
	n1:bignum;
	sum:int64;
procedure dfs(n,from,next:longint);
begin
	if (n>m) then begin
		exit;
	end;
	if (n=m) then begin
		st[tot][0]:=from;
		st[tot][1]:=next;
		inc(tot);
		exit;
	end;
	dfs(n+2,(from<<2)+3,(next<<2)+3);
	dfs(n+1,(from<<1)+1,(next<<1));
	dfs(n+1,(from<<1),(next<<1)+1);
end;
procedure div2(var x:bignum);
var	i,t:longint;
begin
	t:=0;
	//if odd(x[1]) then dec(x[1]);//dec
	for i:=x[0] downto 1 do begin
		x[i]:=t*jw+x[i];
		t:=x[i] mod 2;
		x[i]:=x[i] div 2;
		if x[i]=0 then x[0]:=i-1;
	end;
end;
procedure stom(var s:string;var num:bignum);			//change
var	ts:string[len];
	i:longint;
begin
	fillchar(num,sizeof(num),0);
	ts:='';
	for i:=length(s) downto 1 do begin
		ts:=s[i]+ts;
		if length(ts)=len then begin
			inc(num[0]);
			val(ts,num[num[0]]);
			ts:='';
		end;
	end;
	if ts<>'' then begin
		inc(num[0]);
		val(ts,num[num[0]]);
	end;
end;
function mul(x,y:matrix):matrix;
var     i,j,k,size1:integer;
	z:matrix;
begin
	size1:=(1<<m);
        fillchar(z,sizeof(z),0);
        for i:=0 to size1 do
        for j:=0 to size1 do
	for k:=0 to size1 do begin
		z[i,j]:=(z[i,j]+x[i,k]*y[k,j])mod mo;
	end;
	exit(z);
end;
procedure work(y:bignum);
begin
	fillchar(ans,sizeof(ans),0);
	for i:=0 to (1<<m) do ans[i][i]:=1;
	while (y[0]<>0) do begin
		if odd(y[1]) then begin
			ans:=mul(ans,base);
		end;
		base:=mul(base,base);
		div2(y);
	end;
end;
begin
	readln(s);
	tmp:=pos(' ',s);
	nn:=copy(s,1,tmp-1);
	mm:=copy(s,tmp + 1,length(s) - tmp);val(mm,m);

	tot:=0;
	dfs(0,0,0);{
	if (length(nn)<=2) or (nn='100') then begin
		val(nn,n);
		if odd(n*m) then begin
			writeln(0);
			exit;
		end;	

		dp[0][(1<<m)-1]:=1;
		for i:=1 to n do 
			for j:=0 to tot-1 do begin
				dp[i][st[j][1]]:=(dp[i][st[j][1]]+dp[i-1][st[j][0]]) mod mo;
			end;
		writeln(dp[n][(1<<m)-1]);
	end else begin}
		fillchar(base,sizeof(base),0);
		for j:=0 to tot-1 do begin
			base[st[j][0]][st[j][1]]:=1;
		end;
		stom(nn,n1);
		work(n1);
		sum:=ans[(1<<m)-1][(1<<m)-1];
		writeln(sum mod mo);
	//end;
end.

File: 3473.cpp
#include<stdio.h>
#include<string.h>
int st[3000][2];
int dp[11][3000]; //2^11=2048最多的状态数
int cnt,h,w;
void dfs(int n,int from,int to)//枚举所有一行摆放的可能和其匹配的下层
{
    if(n>w)
        return;
    if(n==w)  //这一行刚好摆完，得到一行摆放的组合
    {
        st[cnt][0]=from;
        st[cnt][1]=to;
        cnt++;
        return;
    }
    dfs(n+2,(from<<2)+3,(to<<2)+3); //横着放，这层和下一层匹配的都多两位1
    dfs(n+1,(from<<1)+1,to<<1);//竖着放，这层多一个1,匹配下层这个位置是0
    dfs(n+1,from<<1,(to<<1)+1);//不放，这层多一个0，匹配的下层多个1

}
int main()
{
    int i,j;
    while(scanf("%d %d",&h,&w)&&h)
    {
        if((w*h)%2)
        {
            printf("0\n");
            continue;
        }
        memset(dp,0,sizeof(dp));
        memset(st,0,sizeof(st));
        cnt=0; //记录每一行摆放所有可能的状态总数
        dfs(0,0,0);
        dp[0][(1<<w)-1]=1;//边界
        for(i=1;i<=h;i++)
        {
            for(j=0;j<cnt;j++)
            {
                dp[i][st[j][1]]+=dp[i-1][st[j][0]];
            }
        }
        printf("%d\n",dp[h][(1<<w)-1]);
    }
    return 0;
}

